### PR TITLE
feat: allow overriding data file with CLI arguments

### PR DESCRIPTION
**Pull request by mspiegel31**
## Description

Users should be able to pass CLI data arguments and use a `--data-file` at the same time.  Let's just take inspiration from the [helm values file](https://helm.sh/docs/chart_template_guide/values_files/) and give highest priority to data values passed at the command line.

from the helm doc site:
>The list above is in order of specificity: values.yaml is the default, which can be overridden by a parent chart's values.yaml, which can in turn be overridden by a user-supplied values file, which can in turn be overridden by --set parameters.

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -199,22 +198,17 @@ def test_data_and_data_file_mutually_exclusive(
...
-        [
-            "Error: Given --data-file, the following are invalid ['--data']" in out,
-            "Error: Given --data, the following are invalid ['--data-file']" in out,
-        ]
-    )
-    assert not (dst / ".copier-answers.yml").exists()
+    assert run_result[1] == 0
+    answers = yaml.safe_load((dst / ".copier-answers.yml").read_text())
+    assert answers["_src_path"] == str(src)
+    assert answers["question"] == expected
```
Comment (yajo) [inline tests/test_cli.py:211]: Could you please also assert that some other question that got answered only by yaml still keeps that value? You'll have to modify a bit more the copier.yml and input.yml files

**Review by yajo — CHANGES_REQUESTED** [review]
Good enhancement, thanks!

File: tests/test_cli.py (hunk 1 above)
Comment (mspiegel31) [reply tests/test_cli.py]: addressed in https://github.com/copier-org/copier/pull/1332/commits/69be4d11e47009e35112ba31067b73dc38010565

**Comment by mspiegel31** [comment]
@yajo I'm a little confused about the coverage failure, it looks like `codecov` is stating this PR reduces test coverage on other test files ([ref](https://app.codecov.io/gh/copier-org/copier/pull/1332/indirect-changes?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=copier-org))?  Would love some clarification or help here.

File: copier/cli.py [hunk 2]
```diff
@@ -103,7 +103,9 @@ class CopierApp(cli.Application):
 class _Subcommand(cli.Application):
     """Base class for Copier subcommands."""
 
-    data: AnyByStrDict = {}
+    def __init__(self, executable):
+        self.data: AnyByStrDict = {}
```
Comment (yajo) [inline copier/cli.py:107]: Why this change?

**Review by yajo — COMMENTED** [review]
Looking good now! 

Just one more doubt...

**Comment by yajo** [comment]
https://github.com/copier-org/copier/pull/1332#issuecomment-1728329806: Some upload error on codecov: https://github.com/copier-org/copier/actions/runs/6251262443/job/16972001045?pr=1332#step:12:51

Gotta check that.

**Comment by yajo** [comment]
I updated the token, hoping (not much) it fixes codecov uploads forever.

File: copier/cli.py (hunk 2 above)
Comment (mspiegel31) [reply copier/cli.py]: when `data` is a _class_ member, test data from previous test runs is unintentionally shared between all tests in the `test_cli.py` test suite.  I've pushed https://github.com/copier-org/copier/pull/1332/commits/ed656c5aabedd414dfa1458d2961c099dde2a21e to demo the CI failure that occurs.  

this failure won't show up if you run just run the failing test on its own (i.e. `pytest tests/test_cli.py -k 'test_data_file_parsed_by_question_type' `); it only shows up when tests before it are ran.

making `data` an _instance_ member solves this issue.  I'm guessing that this only matters for the test suites, since every time you invoke `copier` at the command line it spins up all state "fresh".

**Comment by mspiegel31** [comment]
@yajo/@sisp still not quite sure what to do about these codecov errors;  can you provide any guidance? Given that the "dropped" coverage seems to occur other test files, would you be OK admin merging this one without the `codecov/project` task passing?

File: copier/cli.py (hunk 2 above)
Comment (yajo) [reply copier/cli.py:107]: Ah, mutability... I wish I had the time to rewrite copier in haskell 😂

Thanks!

**Review by yajo — COMMENTED** [review]
Please could you rebase? to see if that fixes codecov.

**Comment by mspiegel31** [comment]
@yajo the rebase fixed this;  should be all set for a final review!

**Review by yajo — APPROVED** [review]
